### ISSUE TITLE
Pull Request test runs failing with Memcache out of memory error

### ISSUE THREAD

**Issue by ElijahLynn** [department-of-veterans-affairs/va.gov-cms#5491]
We are seeing a flicker on pull request test runs which is a frustrating developer experience and a time suck. 

```
Notice: MemcachePool::set(): Server memcache (tcp 11211, udp 0) failed with: SERVER_ERROR out of memory storing o
bject
       (130) in /var/lib/tugboat/docroot/modules/contrib/memcache/src/Driver/MemcacheDriver.php line 17
```

## Acceptance Criteria

- [ ] Issue cause has been investigated & resolved if practicable with limited effort
- [ ] Follow-on issues for investigation & resolution have been created as necessary (AC for resolution issue: failing tests run 20 times consecutively without errors)

**Comment by ElijahLynn** [comment]
We don't see this on Elasticache with test runs, fwiw.

**Comment by ndouglas** [comment]
It occurred to me that since this only started happening recently, that I might be able to get some interesting info from stress-testing two different instances.

For this experiment, I did the following on each of two Tugboat PR instances, one built a couple of weeks ago, one four hours ago:

```bash
# i use nano, all my homies hate vim
sudo apt-get install libevent-dev nano telnet
# build memaslap, a stress-tester that's included in libmemcached
curl -LO https://launchpad.net/libmemcached/1.0/1.0.18/+download/libmemcached-1.0.18.tar.gz
tar xzf libmemcached-1.0.18.tar.gz
cd libmemcached-1.0.18
./configure --enable-memaslap
nano -w ./Makefile 
# find "CXXFLAGS = " and append "-fpermissive" to avoid a compiler error
# find "LDFLAGS = " and append "-L/lib64 -lpthread" bc pthreads
make
```

Then I did 

```bash
./clients/memaslap -s memcache:11211 -t 360s
```

On the older instance, `608d5ac3eac8ecfcf8a7212d`, I received the following output:

```
servers : memcache:11211
threads count: 1
concurrency: 16
run time: 360s
windows size: 10k
set proportion: set_prop=0.10
get proportion: get_prop=0.90
cmd_get: 36318242
cmd_set: 4035369
get_misses: 1461599
written_bytes: 7001356495
read_bytes: 38730449121
object_bytes: 4390481472

Run time: 360.0s Ops: 40353611 TPS: 112084 Net_rate: 121.1M/s
```

On the newer instance, `60af7d694fa3233d81cf1832`, I received the following output:

```
servers : memcache:11211
threads count: 1
concurrency: 16
run time: 360s
windows size: 10k
set proportion: set_prop=0.10
get proportion: get_prop=0.90
<12 SERVER_ERROR out of memory storing object
<18 SERVER_ERROR out of memory storing object
<9 SERVER_ERROR out of memory storing object
<15 SERVER_ERROR out of memory storing object
<12 SERVER_ERROR out of memory storing object
... + ~1000 more lines of this, interestingly all displayed in the first few seconds of the run ...
cmd_get: 36840689
cmd_set: 4094475
get_misses: 9878221
written_bytes: 7103240950
read_bytes: 30010506722
object_bytes: 4454788800

Run time: 360.0s Ops: 40935164 TPS: 113699 Net_rate: 98.3M/s
```

This would seem to indicate that there's a pathological condition in the more recently built environment that does not exist in the older environment.

**Comment by ndouglas** [comment]
There are some interesting differences in stats between the older (passing) and newer (failing) environments:

```bash
paste ~/a.txt ~/b.txt | awk '{ printf "%40s\t%20s\t%20s\n",$2,$3,$6 }'
                                     pid	                   7	                   7
                                  uptime	               19825	                5747
                                    time	          1622757114	          1622757095
                                 version	               1.6.9	               1.6.9
                                libevent	        2.1.8-stable	        2.1.8-stable
                            pointer_size	                  64	                  64
                             rusage_user	          168.439964	          254.216695
                           rusage_system	          576.320724	          915.587713
                         max_connections	                1024	                1024
                        curr_connections	                   2	                   2
                       total_connections	                4727	                 337
                    rejected_connections	                   0	                   0
                   connection_structures	                  68	                  20
                        response_obj_oom	                   0	                   0
                      response_obj_count	                   1	                   1
                      response_obj_bytes	               65536	               65536
                          read_buf_count	                  24	                  12
                          read_buf_bytes	              393216	              196608
                     read_buf_bytes_free	              311296	              114688
                            read_buf_oom	                   0	                   0
                            reserved_fds	                  20	                  20
                                 cmd_get	            43505253	            71241958
                                 cmd_set	             4740350	             7915922
                               cmd_flush	                   0	                   0
                               cmd_touch	                   0	                   0
                                cmd_meta	                   0	                   0
                                get_hits	            28020798	            68338157
                              get_misses	            15484455	             2903801
                             get_expired	                   0	                   0
                             get_flushed	                   0	                   0
                           delete_misses	                4960	                 754
                             delete_hits	                3129	                 374
                             incr_misses	                   0	                   0
                               incr_hits	                   0	                   0
                             decr_misses	                   0	                   0
                               decr_hits	                   0	                   0
                              cas_misses	                   0	                   0
                                cas_hits	                   0	                   0
                              cas_badval	                   0	                   0
                              touch_hits	                   0	                   0
                            touch_misses	                   0	                   0
                               auth_cmds	                   0	                   0
                             auth_errors	                   0	                   0
                              bytes_read	          8839769874	         13739893818
                           bytes_written	         33349914855	         75936018154
                          limit_maxbytes	           268435456	           268435456
                         accepting_conns	                   1	                   1
                     listen_disabled_num	                   0	                   0
              time_in_listen_disabled_us	                   0	                   0
                                 threads	                   4	                   4
                             conn_yields	                  11	                   0
                        hash_power_level	                  17	                  18
                              hash_bytes	             1048576	             2097152
                       hash_is_expanding	                   0	                   0
                   slab_reassign_rescues	                1053	                   0
             slab_reassign_chunk_rescues	                   0	                   0
           slab_reassign_evictions_nomem	               55603	                   0
            slab_reassign_inline_reclaim	                  71	                   0
                slab_reassign_busy_items	                   5	                   0
              slab_reassign_busy_deletes	                   0	                   0
                   slab_reassign_running	                   0	                   0
                             slabs_moved	                 207	                   0
                     lru_crawler_running	                   0	                   0
                      lru_crawler_starts	                  26	                  14
                  lru_maintainer_juggles	             1863252	             1658733
                            malloc_fails	                   0	                   0
                      log_worker_dropped	                   0	                   0
                      log_worker_written	                   0	                   0
                     log_watcher_skipped	                   0	                   0
                        log_watcher_sent	                   0	                   0
                     unexpected_napi_ids	                   0	                   0
                    round_robin_fallback	                   0	                   0
                                   bytes	           244507969	           246779294
                              curr_items	              192546	              215082
                             total_items	             4741403	             7915922
                   slab_global_page_pool	                   0	                   0
                       expired_unfetched	                   0	                   0
                       evicted_unfetched	              640430	              122476
                          evicted_active	                 647	                 195
                               evictions	             4396738	             7700459
                               reclaimed	                   0	                   0
                       crawler_reclaimed	                   0	                   0
                   crawler_items_checked	             1984877	             1751154
                       lrutail_reflocked	                2247	                 969
                           moves_to_cold	            14587989	            29622648
                           moves_to_warm	            10739545	            23907921
                        moves_within_lru	             9054204	            23723677
                         direct_reclaims	             4423056	             7700459
                       lru_bumps_dropped	                   0	                   0
```

These lines in particular interest me:

```
                   slab_reassign_rescues	                1053	                   0
             slab_reassign_chunk_rescues	                   0	                   0
           slab_reassign_evictions_nomem	               55603	                   0
            slab_reassign_inline_reclaim	                  71	                   0
                slab_reassign_busy_items	                   5	                   0
              slab_reassign_busy_deletes	                   0	                   0
                   slab_reassign_running	                   0	                   0
                             slabs_moved	                 207	                   0
```

**Comment by ndouglas** [comment]
I switched from `memaslap` to [`mc-crusher`](https://github.com/memcached/mc-crusher), because it's written by the `memcached` team and not the `libmemcached` team.  It's easy to compile on Tugboat -- just requires `libevent-dev`.

The command can be run like this:

```bash
./mc-crusher --conf ./conf/<config file> --ip memcache
```

Most of these are uneventful, as observed by `memcached-tool-ng` in another shell.  However, running

```bash
./mc-crusher --conf ./conf/slab_rebal_torture --ip memcache
```

and 

```bash
./mc-crusher --conf ./conf/slab_rebal_torture2 --ip memcache
```

result in rapidly escalating OOMs, according to `memcached-tool-ng` (these aren't displayed by `mc-crusher`, and I'm not sure why).

```
memcache:11211:
   #  Item_Size    Max_age   Pages     Count   Used Mem     To Store   Efficiency              Evicted  Evict_Time          OOM
   1        96B         3s       2     21844    2.00 MB      1.47 MB      73.51 %               482107           3          785
   2       120B         3s       3     26214    3.00 MB      2.49 MB      83.10 %               557086           3           88
   3       152B         3s       3     20694    3.00 MB      2.76 MB      92.02 %               512618           3          437
   4       192B         4s       5     23612    5.00 MB      3.82 MB      76.50 %               515435           3          300
   5       240B         2s       6     26214    6.00 MB      5.49 MB      91.52 %               420327           2          255
   6       304B         2s       8     27591    8.00 MB      7.10 MB      88.72 %               327168           2          775
   7       384B        40s      12     30003   12.00 MB     10.58 MB      88.13 %               241860           9            0
   8       480B         2s      12     25000   12.00 MB     11.20 MB      93.29 %               313005           2          527
   9       600B        40s      12     20001   12.00 MB     10.86 MB      90.51 %               325272           2          208
  10       752B         1s      11     15000   11.00 MB      9.59 MB      87.16 %               362758           1          413
  11       944B        40s      10     10004   10.00 MB      8.30 MB      82.99 %               227541           1           49
  12       1.2K      1316s     158    139829  158.00 MB    130.66 MB      82.70 %              2380702        1316            0
  13       1.4K         0s       8      5000    8.00 MB      6.53 MB      81.65 %                57742           1            0
  14       1.8K         0s       5      2500    5.00 MB      4.22 MB      84.38 %                81812           0            0
  15       2.3K        40s       1         1    1.00 MB      2.04 KB       0.20 %                    0           0            0
  19       5.5K        40s       1         2    1.00 MB     10.26 KB       1.00 %                    0           0            0
  20       6.9K        40s       1         1    1.00 MB      6.10 KB       0.60 %                    0           0            0
  23      13.6K        40s       1         1    1.00 MB     12.77 KB       1.25 %                    0           0            0
  26      26.5K        40s       1         1    1.00 MB     22.31 KB       2.18 %                    0           0            0
  29      51.7K        40s       1         1    1.00 MB     47.46 KB       4.64 %                    0           0            0
```

I'm going to shell into one of the BRD environments and see if this behaves similarly there.

**Comment by ndouglas** [comment]
differential config (Elasticache vs. Tugboat vs. Lando):

```
            maxbytes     13932429312       268435456        67108864
launch_time_maxbytes     13932429312            1024            1024
            maxconns           65000           11211           11211
             tcpport           11211               0               0
             udpport           11211            NULL            NULL
               inter            NULL               0               2
           verbosity               1               0               0
              oldest         1181266              on              on
           evictions              on            NULL            NULL
       domain_socket            NULL             700             700
               umask             700            1.25            1.25
       growth_factor            1.25              48              48
          chunk_size              48               4               4
         num_threads               2               4               4
 num_threads_per_udp               2               :               :
     stat_key_prefix               :              no              no
      detail_enabled              no              20              20
      reqs_per_event              20             yes             yes
         cas_enabled             yes            1024            1024
         tcp_backlog            1024  auto-negotiate  auto-negotiate
    binding_protocol  auto-negotiate              no              no
   auth_enabled_sasl              no              no              no
  auth_enabled_ascii              no         1048576         1048576
       item_size_max         1048576             yes             yes
          config_max              16               0               0
     config_size_max           65536             yes             yes
       maxconns_fast             yes               1               1
      hashpower_init              16            0.80            0.80
       slab_reassign             yes              30              30
       slab_automove               0          524288          524288
 slab_automove_ratio            0.80             yes             yes
slab_automove_window              30             100             100
      slab_chunk_max          524288               0               0
         lru_crawler             yes               0               0
   lru_crawler_sleep             100             yes             yes
 lru_crawler_tocrawl               0             yes             yes
    tail_repair_time               0         murmur3         murmur3
       flush_enabled             yes             yes             yes
        dump_enabled             yes             yes             yes
      hash_algorithm         jenkins              20              20
lru_maintainer_thread             yes              40              40
       lru_segmented             yes            0.20            0.20
         hot_lru_pct              20            2.00            2.00
        warm_lru_pct              40              no              no
      hot_max_factor            0.20              61              61
     warm_max_factor            2.00               0               0
            temp_lru              no          262144          262144
       temporary_ttl              61           65536           65536
        idle_timeout               0               0               0
 watcher_logbuf_size          262144              no               0
  worker_logbuf_size           65536              no              no
  resp_obj_mem_limit               0             512              no
  read_buf_mem_limit               0      4294967295             512
         track_sizes              no               0      4294967295
inline_ascii_response              no            2000               0
       ext_item_size             512         4194304            2000
        ext_item_age      4294967295               0         4194304
         ext_low_ttl               0               0               0
    ext_recache_rate            2000            0.80               0
       ext_wbuf_size         4194304           0.010            0.80
   ext_compact_under               0              no           0.010
      ext_drop_under               0              no              no
        ext_max_frag            0.80          (null)                
slab_automove_freeratio           0.010          (null)                
     ext_drop_unread              no               0                
       ssl_keyformat               1                                
         ssl_ciphers            NULL                                
         ssl_ca_cert            NULL                                
       ssl_wbuf_size           16384                                
   ssl_session_cache              no                                
        num_napi_ids          (null)                                
         memory_file          (null)                                
```

**Comment by ElijahLynn** [comment]
K, @ndouglas and I just figured out the solution with the help of @dormando!

The issue is better outlined in https://github.com/memcached/memcached/issues/793#issuecomment-857112239 so I'll just put the TLDR of the solution here for now:

We extended the Tugboat Memcached Docker image and put it in https://github.com/va-cms-bot/docker-memcached-tugboat and uploaded the image to https://hub.docker.com/r/vacmsbot/docker-memcached-tugboat/tags (1.6.6 tag, same as Elasticache). 

I did 15 PRs and none of them failed with "out of memory" anymore. 
![image](https://user-images.githubusercontent.com/1504756/121271815-5655e580-c879-11eb-97e6-2a72a3600d67.png)
